Sources/InfomaniakCore/SuddenTermination/ExpiringActivity.swift

```diff
@@ -0,0 +1,129 @@
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Foundation
+
+/// Delegation mechanism to notify the end of an `ExpiringActivity`
+public protocol ExpiringActivityDelegate: AnyObject {
+    /// Called when the system is requiring us to terminate an expiring activity
+    func backgroundActivityExpiring()
+}
+
+/// Something that can perform arbitrary short background tasks, with a super simple API.
+public protocol ExpiringActivityable {
+    /// Common init method
+    /// - Parameters:
+    ///   - id: Something to identify the background activity in debug
+    ///   - qos: QoS used by the underlying queues
+    ///   - delegate: The delegate to notify we should terminate
+    init(id: String, qos: DispatchQoS, delegate: ExpiringActivityDelegate?)
+
+    /// init method
+    /// - Parameters:
+    ///   - id: Something to identify the background activity in debug
+    ///   - delegate: The delegate to notify we should terminate
+    init(id: String, delegate: ExpiringActivityDelegate?)
+
+    /// Register with the system an expiring activity
+    func start()
+
+    /// Terminate all the expiring activities
+    func endAll()
+
+    /// True if the system asked to stop the background activity
+    var shouldTerminate: Bool { get }
+}
+
+public final class ExpiringActivity: ExpiringActivityable {
+    private let qos: DispatchQoS
+
+    private let queue: DispatchQueue
+
+    private let processInfo = ProcessInfo.processInfo
+
+    var locks = [TolerantDispatchGroup]()
+
+    let id: String
+
+    public var shouldTerminate = false
+
+    weak var delegate: ExpiringActivityDelegate?
+
+    // MARK: Lifecycle
+
+    public init(id: String, qos: DispatchQoS, delegate: ExpiringActivityDelegate?) {
+        self.id = id
+        self.qos = qos
+        self.delegate = delegate
+        queue = DispatchQueue(label: "com.infomaniak.ExpiringActivity.sync", qos: qos)
+    }
+
+    public convenience init(id: String = "\(#function)-\(UUID().uuidString)", delegate: ExpiringActivityDelegate? = nil) {
+        self.init(id: id, qos: .userInitiated, delegate: delegate)
+    }
+
+    deinit {
+        queue.sync {
+            assert(locks.isEmpty, "please make sure to call 'endAll()' once explicitly before releasing this object")
+        }
+    }
+
+    public func start() {
+        let group = TolerantDispatchGroup(qos: qos)
+
+        queue.sync {
+            self.locks.append(group)
+        }
+
+        #if os(macOS)
+        // We block a non cooperative queue that matches current QoS
+        DispatchQueue.global(qos: qos.qosClass).async {
+            self.processInfo.performActivity(options: .suddenTerminationDisabled, reason: self.id) {
+                // No expiration handler as we are running on macOS
+                group.enter()
+                group.wait()
+            }
+        }
```

Comment:
Why is a DispatchQueue needed on macOS but not on iOS ?
IS `performActivity` is synchronous on macOS ?

Reply:
> Is performActivity synchronous on macOS ?
Talking from memory, that is the behaviour I had on my machine.
Mac and iOS do behave differently.

This code is in prod and works, but lives within the app.

Reply:
I will create a test like discussed.